src/Toml/Parser.hs

```diff
@@ -258,10 +259,25 @@ dateTimeP = lexeme $ try hoursP <|> dayLocalZoned
         Just frc' -> return (rdPico $ int ++ "." ++ frc')
 
 arrayP :: Parser [UValue]
-arrayP = lexeme $ between (char '[' *> sc) (char ']') elements
+arrayP = lexeme (between (char '[' *> sc) (char ']') elements) <?> "array"
   where
     elements :: Parser [UValue]
-    elements = valueP `sepEndBy` spComma <* skipMany (text ",")
+    elements = do v <- valueP -- Parse the first value to determine the type
+                  sep <- optional spComma
+                  vs <- case sep of
+                          Nothing -> return []
+                          Just _  -> (element v `sepEndBy` spComma) <* skipMany spComma
+                  return (v:vs)
+               <|> return []
+
+    element :: UValue -> Parser UValue
+    element = \case
+                 UBool    _ -> UBool    <$> boolP
```

Comment:
I think that indentation here is too big... 4 spaces should be enough. But I like that you've used `LambdaCase` extension here and aligned `->` under each other! :+1: